Adding the rearrangement most important configuration keys to the documentation (#1126)
facebookresearch/habitat-lab: habitat-lab/habitat/config/CONFIG_KEYS.md

# Configuration Keys

## Dataset
Configuration for the dataset of the task.
A dataset consists of episodes
(a start configuration for a task within a scene) and a scene dataset
(with all the assets needed to instantiate the task)

| Key | Description |
| --- | --- |
| habitat.dataset.type |  The key for the dataset class that will be used. Examples of such keys are `PointNav-v1`, `ObjectNav-v1`, `InstanceImageNav-v1` or `RearrangeDataset-v0`. Different datasets have different properties so you should use the dataset that fits your task. |
| habitat.dataset.scene_dir | The path to the directory containing the scenes that will be used. You should put all your scenes in the same folder (example `data/scene_datasets`) to avoid having to change it. |
|habitat.dataset.data_path | The path to the episode dataset. Episodes need to be compatible with the `type` argument (so they will load properly) and only use scenes that are present in the `scenes_dir`.|

> |habitat.dataset.split | `data_path` can have a `split` in the path. For example: "data/datasets/pointnav/habitat-test-scenes/v1/{split}/{split}.json.gz" the value in "{split}" will be replaced by the value of the `split` argument. This allows to easily swap between `train` and `eval` episodes by only changing the split argument.|

Author: The splits are present in the dataset. The value of split used depends on how the dataset is made. There is no other way than just looking into the dataset files and checking what are the available splits.
Documenting how to create datasets and splits is not part of this PR.

## Task
The definition of the task in Habitat.
There are many different Tasks determined by the `habitat.task.type` config:
 - Point navigation : `Nav-0`

>  - Image navigation : `Nav-0`

Author: Both [pointnav](https://github.com/facebookresearch/habitat-lab/blob/main/habitat-lab/habitat/config/habitat/task/pointnav.yaml) and [imagenav](https://github.com/facebookresearch/habitat-lab/blob/main/habitat-lab/habitat/config/habitat/task/imagenav.yaml) use Nav-0 for their task. I guess task got reused because the two are not that different.

| habitat.task.base_angle| For Rearrangement tasks only. Controls the standard deviation of the random normal noise applied to the base's position at the start of an episode.|

## Visual Agents
You can inherit a default agent in your configuration.
The some sample agents are defined [here](./habitat/simulator/agents/) and are reused in many of our benchmarks. To use one of these agents, add
```
  - /habitat/simulator/agents@habitat.simulator.agents.main_agent: <the key of the agent>
```
to your defaults list. All agents have a camera resolution of 256 by 256.
| Key | Description |
| --- | --- |
|depth_head_agent| An Agent with a single depth camera attached to its head.|
|rgb_head_agent| An Agent with a single RGB camera attached to its head.|
|rgbd_head_rgbd_arm_agent| An Agent with both a RGB and depth camera attached to its head and arm.|
|depth_head_agent_vis| An Agent with a depth camera attached to its head and a third person RGB camera on top of the robot.|

## Environment
Some habitat environment configurations.
| Key | Description |
| --- | --- |
|habitat.environment.max_episode_steps| The maximum number of environment steps before the episode ends.|

> |habitat.environment.max_episode_seconds| The maximum number of seconds steps before the episode ends.|

Reviewer: is this seconds in wall-clock time?

> ## Navigation Actions

Reviewer: Maybe we explicitly call this section "Discrete navigation Measures" since there are also navigation actions in rearrangement but continuous?

> The way one would add an action to a configuration file would be by adding to the `defaults` list. For example:

Reviewer: can you describe what actions are? Typically, in RL actions are defined by the env, for example if actions are joint angles, torques, end-effector position.
Is this what we are trying to define here? So is this actions dictionary meant to be added to a task config?

| habitat.task.actions.empty | In Navigation tasks only, the pass action. The robot will do nothing.|
| habitat.task.actions.move_forward |     In Navigation tasks only, this discrete action will move the robot forward by a fixed amount determined by the `habitat.simulator.forward_step_size` amount. |
| habitat.task.actions.turn_left |     In Navigation tasks only, this discrete action will rotate the robot to the left  by a fixed amount determined by the `habitat.simulator.turn_angle` amount. |
| habitat.task.actions.turn_right |     In Navigation tasks only, this discrete action will rotate the robot to the right by a fixed amount determined by the `habitat.simulator.turn_angle` amount. |
| habitat.task.actions.look_up |      In Navigation tasks only, this discrete action will rotate the robot's camera up by a fixed amount determined by the `habitat.simulator.tilt_angle` amount. |
| habitat.task.actions.look_down |      In Navigation tasks only, this discrete action will rotate the robot's camera down by a fixed amount determined by the `habitat.simulator.tilt_angle` amount. |

> ## Navigation Measures

Reviewer: since these measures hold for both navigation and rearrangement, we should rename this to be just Measures

Reviewer: Can you please describe what measurement are in one sentence?
Maybe something like " measurements are (sometimes privileged) quantities that we extract from the simulation, that can be used for calculating rewards, success metrics, etc."

> The way one would add an action to a configuration file would be by adding to the `defaults` list. For example:

Reviewer: ```suggestion
The way one would add a measurement to a configuration file would be by adding to the `defaults` list. For example:
```

```
defaults:
  - /habitat/task/measurements:
    - robot_force
    - force_terminate
```

| Key | Description |
| --- | --- |
| habitat.task.measurements.num_steps| In both Navigation and Rearrangement tasks, counts the number of steps since  the start of the episode.|

> | habitat.task.measurements.distance_to_goal | In Navigation tasks only, measures the geodesic distance to the goal.|

Reviewer: why only in navigation? Can't we use this in rearrangement, for example, for distance to object?

Author: Nope. Different measure.

> |habitat.task.measurements.distance_to_goal.distance_to | If 'POINT' measures the distance to the closest episode goal. If 'VIEW_POINTS' measures the distance to the episode's goal's viewpoint. |

Reviewer: sorry I don't understand this one. if you could explain it to me, maybe we can re-write it?

Author: Sure. This is the code : 
```
            if self._distance_to == "POINT":
                distance_to_target = self._sim.geodesic_distance(
                    current_position,
                    [goal.position for goal in episode.goals],
                    episode,
                )
            elif self._distance_to == "VIEW_POINTS":
                distance_to_target = self._sim.geodesic_distance(
                    current_position, self._episode_view_points, episode
                )
            else:
                logger.error(
                    f"Non valid distance_to parameter was provided: {self._distance_to }"
                )

```
so if this config key us set to POINT, the distance is relative to the closest goal position. if it is VIEW_POINT, then the distance is relative to the closest view point. I **ASSUME** this is the difference between navigating close to an object and navigating to the POV from which an image of the object is taken.

|habitat.task.measurements.success.success_distance| The minimal distance the robot must be to the goal for a success.|
|habitat.task.measurements.spl|    For Navigation tasks only, Measures the SPL (Success weighted by Path Length) ref: [On Evaluation of Embodied Agents - Anderson et. al](https://arxiv.org/pdf/1807.06757.pdf)  Measure is always 0 except at success where it will be  the ratio of the optimal distance from start to goal over the total distance  traveled by the agent. Maximum value is 1. `SPL = success * optimal_distance_to_goal / distance_traveled_so_far`
|habitat.task.measurements.soft_spl |     For Navigation tasks only, Similar to SPL, but instead of a boolean, success is now calculated as 1 - (ratio of distance covered to target).   `SoftSPL = max(0, 1 - distance_to_goal / optimal_distance_to_goal) * optimal_distance_to_goal / distance_traveled_so_far`
|habitat.task.measurements.distance_to_goal_reward    |    In Navigation tasks only, measures a reward based on the distance towards the goal. The reward is `- (new_distance - previous_distance)` i.e. the decrease of distance to the goal.

## Navigation Lab Sensors

> Lab sensors are any non-rendered sensor observation. Like geometric goal information. The way one would add an action to a configuration file would be by adding to the `defaults` list. For example:

Reviewer: ```suggestion
Lab sensors are any non-rendered sensor observation, like geometric goal information. The way one would add an action to a configuration file would be by adding to the `defaults` list. For example:
```

```
defaults:
  - /habitat/task/lab_sensors:
    - objectgoal_sensor
    - compass_sensor
```

| Key | Description |
| --- | --- |

> |  habitat.task.lab_sensors.objectgoal_sensor|  For Object Navigation tasks only. Generates a discrete observation containing the id of the goal object for the episode. |

Reviewer: isn't this also used in rearrangement tasks?

Author: Nope. Navigation only. It needs a specific field in the episode.

> |  habitat.task.lab_sensors.objectgoal_sensor.goal_spec| A string that can take the value TASK_CATEGORY_ID or OBJECT_ID. If the value is TASK_CATEGORY_ID, then the observation will be the id of the `episode.object_category` attribute, if the value is OBJECT_ID, then the observation will be the id of the first goal object. |

Reviewer: can you also briefly explain what these ids do? Are these somehow the goal specification for the task? I can imagine that it is either the object category or the specific object id, but a short clarification would be helpful.

|  habitat.task.lab_sensors.instance_imagegoal_sensor  |    Used only by the InstanceImageGoal Navigation task. The observation is a rendered image of the goal object within the scene.|
|  habitat.task.lab_sensors. instance_imagegoal_hfov_sensor |     Used only by the InstanceImageGoal Navigation task. The observation is a single float value corresponding to the Horizontal field of view (HFOV) in degrees of  the image provided by the `habitat.task.lab_sensors.instance_imagegoal_sensor `.|
|  habitat.task.lab_sensors.compass_sensor |     For Navigation tasks only. The observation of the `EpisodicCompassSensor` is a single float value corresponding to the angle difference in radians between the current rotation of the robot and the start rotation of the robot along the vertical axis. |
|  habitat.task.lab_sensors.gps_sensor |     For Navigation tasks only. The observation of the EpisodicGPSSensor are two float values corresponding to the vector difference in the horizontal plane between the current position and the start position of the robot (in meters). |

## Rearrangement Action

| Key | Description |
| --- | --- |

> |habitat.task.actions.arm_action |In Rearrangement tasks only, the action that will move the robot arm around. |

Reviewer: does this have to be joint angles? or velocities/torques are also allowed?

Author: I haven't found an actuator that does that. arm_action is the only actuator I found that seem to be able to move the arm. And even if it is not, it is the only one we use in our benchmarks.

Reviewer: I just mean can we say what is the action? Is it joint angles? If yes, can we say
"In Rearrangement tasks only, the desired joint angles for the robot arm. Useful for moving the arm."

|habitat.task.actions.arm_action.grip_controller| Can either be None,  `MagicGraspAction` or `SuctionGraspAction`. If None, the arm will be unable to grip object. Magic grasp will grasp the object if the end effector is within grasp_thresh_dist of an object, with `SuctionGraspAction`, the object needs to be in contact with the end effector. |
|habitat.task.actions.base_velocity |     In Rearrangement only. Corresponds to the base velocity. Contains two continuous actions, the first one controls forward and backward motion, the second the rotation.
|habitat.task.actions.rearrange_stop     | In rearrangement tasks only, if the robot calls this action, the task will end.|
|habitat.task.actions.oracle_nav_action| Rearrangement Only, Oracle navigation action. This action takes as input a discrete ID which refers to an object in the PDDL domain. The oracle navigation controller then computes the actions to navigate to that desired object.|

## Rearrangement Sensors

| Key | Description |
| --- | --- |

> |habitat.task.lab_sensors.relative_resting_pos_sensor     | Rearrangement only. Sensor for the relative position of the end-effector's resting position, relative to the end-effector's current position. The three values correspond to the cartesian coordinates of the resting position in the frame of reference of the end effector. The desired resting position is determined by the habitat.task.desired_resting_position coordinates relative to the robot's base.|

Reviewer: ```suggestion
|habitat.task.lab_sensors.relative_resting_pos_sensor     | Rearrangement only. Sensor for the desired position of the end-effector's resting position, relative to the end-effector's current position. The three values correspond to the cartesian coordinates of the desired resting position in the frame of reference of the current end effector. The desired resting position is determined by the habitat.task.desired_resting_position coordinates relative to the robot's base.|
```

| habitat.task.lab_sensors.is_holding_sensor | Rearrangement only. A single float sensor with value 1.0 if the robot is grasping any object and 0.0 otherwise.|
|habitat.task.lab_sensors.end_effector_sensor | Rearrangement only. the cartesian coordinates (3 floats) of the arm's end effector in the frame of reference of the robot's base.|
|habitat.task.lab_sensors.joint_sensor| Rearrangement only. Returns the joint positions of the robot.|

> | habitat.task.lab_sensors.goal_sensor |     Rearrangement only. Returns the relative position from end effector to a goal position in which the agent needs to place an object. |

Reviewer: what happens if not holding an object?

Author: The same thing I guess:

```
    def get_observation(self, observations, episode, *args, **kwargs):
        global_T = self._sim.get_robot_data(self.robot_id).robot.ee_transform
        T_inv = global_T.inverted()

        _, pos = self._sim.get_targets()
        return batch_transform_point(pos, T_inv, np.float32).reshape(-1)
        ```

> | habitat.task.lab_sensors.target_start_gps_compass_sensor |     Rearrangement only. Returns the initial position of every object that needs to be rearranged in composite tasks, in 2D polar coordinates. |

Reviewer: ```suggestion
| habitat.task.lab_sensors.target_start_gps_compass_sensor |     Rearrangement only. Returns the initial position of every object that needs to be rearranged in composite tasks relative to the robot's start position, in 2D polar coordinates. |
```

> | habitat.task.lab_sensors.target_goal_gps_compass_sensor |    Rearrangement only. Returns the desired goal position of every object that needs to be rearranged in composite tasks, in 2D polar coordinates. |

Reviewer: ```suggestion
| habitat.task.lab_sensors.target_goal_gps_compass_sensor |    Rearrangement only. Returns the desired goal position of every object that needs to be rearranged in composite tasks relative to the robot's start position, in 2D polar coordinates. |
```

> | habitat.task.lab_sensors.target_goal_gps_compass_sensor |    Rearrangement only. Returns the desired goal position of every object that needs to be rearranged in composite tasks, in 2D polar coordinates. |

Reviewer: is this 2D polar space described somewhere?

## Rearrangement Measures
| Key | Description |
| --- | --- |

> |habitat.task.measurements.end_effector_to_rest_distance | Rearrangement only. Distance between current end effector position  and the resting position of the end effector. Requires that the   RelativeRestingPositionSensor is attached to the agent. |

Reviewer: how do we attach a sensor to an agent? (can we point to an example code)

Author: We have it in the Navigation sensors. I feel it would be duplication.

> |habitat.task.measurements.robot_force |      The amount of force in newton's applied by the robot. It computes both the instant and accumulated. |

Reviewer: ```suggestion
|habitat.task.measurements.robot_force |      The amount of force in newton's applied by the robot. It computes both the instantaneous and accumulated force during the episode. |
```

|habitat.task.measurements.does_want_terminate | Rearrangement Only. Measures 1 if the agent has called the stop action and 0 otherwise.    |
|habitat.task.measurements.force_terminate |    If the force is greater than a certain threshold, this measure will be 1.0 and 0.0 otherwise.   Note that if the measure is 1.0, the task will end as a result. |

> |habitat.task.measurements.force_terminate.max_accum_force |  The threshold for the accumulated force. -1 is no threshold.   |

Reviewer: ```suggestion
|habitat.task.measurements.force_terminate.max_accum_force |  The threshold for the accumulated force before calling termination. -1 is no threshold, i.e., force-based termination is never called.   |
```

> |habitat.task.measurements.force_terminate.max_instant_force |  The threshold for the current, instant force. -1 is no threshold.   |

Reviewer: ```suggestion
|habitat.task.measurements.force_terminate.max_instant_force |  The threshold for the current, instantaneous force before calling termination. -1 is no threshold, i.e., force-based termination is never called.   |
```

|habitat.task.measurements.object_to_goal_distance |  In rearrangement only. The distance between the target object and the goal position for the object. |
|habitat.task.measurements.obj_at_goal |  The measure is a dictionary of target indexes to float. The values are 1 if the object is within succ_thresh of the goal position for that object.   |
|habitat.task.measurements.obj_at_goal.succ_thresh | The threshold distance below which an object is considered at the goal location.    |

> |habitat.task.measurements.art_obj_at_desired_state |   Rearrangement open/close container tasks only. Whether the articulated object (fridge or cabinet door) towards a desired state (open or closed) as defined by the task. |

Reviewer: ```suggestion
|habitat.task.measurements.art_obj_at_desired_state |   Rearrangement open/close container tasks only. Whether the articulated object (fridge or cabinet door) is in a desired state (open or closed) as defined by the task. |
```

Reviewer: How is open/close measured? based on articulated joint angles?

Author: There is a target articulated object joint state in the episode. The code seems to indicate this is based of joint state.

> |habitat.task.measurements.distance_to_goal.distance_to | If 'POINT' measures the distance to the closest episode goal. If 'VIEW_POINTS' measures the distance to the episode's goal's viewpoint. |

Reviewer: ```suggestion
|habitat.task.measurements.distance_to_goal.distance_to | If 'POINT' measures the distance to the closest episode goal. If 'VIEW_POINTS' measures the distance to the episode goal viewpoints (useful in image nav). |
```

|habitat.task.lab_sensors.end_effector_sensor | Rearrangement only. the cartesian coordinates (3 floats) of the arm's end effector in the frame of reference of the robot's base.|
|habitat.task.lab_sensors.joint_sensor| Rearrangement only. Returns the joint positions of the robot.|
| habitat.task.lab_sensors.goal_sensor |     Rearrangement only. Returns the relative position from end effector to a goal position in which the agent needs to place an object. |
| habitat.task.lab_sensors.target_start_gps_compass_sensor |     Rearrangement only. Returns the initial position of every object that needs to be rearranged in composite tasks relative to the robot's start position, in 2D polar coordinates (distance and angle in the horizontal plane). |
| habitat.task.lab_sensors.target_goal_gps_compass_sensor |    Rearrangement only. Returns the desired goal position of every object that needs to be rearranged in composite tasks relative to the robot's start position, in 2D polar coordinates (distance and angle in the horizontal plane). |

## Rearrangement Measures
| Key | Description |
| --- | --- |

> |habitat.task.measurements.end_effector_to_rest_distance | Rearrangement only. Distance between current end effector position  and the resting position of the end effector. Requires that the   RelativeRestingPositionSensor is attached to the agent. |

Reviewer: ```suggestion
|habitat.task.measurements.end_effector_to_rest_distance | Rearrangement only. Distance between current end effector position  and the resting position of the end effector. Requires that the   RelativeRestingPositionSensor is attached to the agent (see Rearragement Sensors above to see how to attach sensors). |
```
